src/arkode/arkode_lsrkstep.c

```diff
@@ -0,0 +1,2232 @@
+  present.  If any of them is missing it returns SUNFALSE.
+  ---------------------------------------------------------------*/
+sunbooleantype lsrkStep_CheckNVector(N_Vector tmpl)
+{
+  if ((tmpl->ops->nvclone == NULL) || (tmpl->ops->nvdestroy == NULL) ||
+      (tmpl->ops->nvlinearsum == NULL) || (tmpl->ops->nvconst == NULL) ||
+      (tmpl->ops->nvscale == NULL) || (tmpl->ops->nvwrmsnorm == NULL) ||
+      (tmpl->ops->nvspace == NULL))
+  {
+    return SUNFALSE;
+  }
+  return SUNTRUE;
+}
+
+/*---------------------------------------------------------------
+  lsrkStep_DomEigUpdateLogic:
+
+  This routine checks if the step is accepted or not and reassigns
+  the dom_eig update flags accordingly.
+  ---------------------------------------------------------------*/
+
+void lsrkStep_DomEigUpdateLogic(ARKodeMem ark_mem, ARKodeLSRKStepMem step_mem,
+                                sunrealtype dsm)
+{
+  if (dsm <= ONE)
+  {
+    N_VScale(ONE, ark_mem->tempv2, ark_mem->fn);
+    ark_mem->fn_is_current = SUNTRUE;
+
+    step_mem->dom_eig_is_current = (step_mem->const_Jac == SUNTRUE);
+
+    step_mem->dom_eig_update = SUNFALSE;
+    if (step_mem->dom_eig_nst + step_mem->dom_eig_freq <=
+        ark_mem->nst_attempts + 1)
+    {
+      step_mem->dom_eig_update = !step_mem->dom_eig_is_current;
+    }
+  }
+  else { step_mem->dom_eig_update = !step_mem->dom_eig_is_current; }
+}
+
+/*---------------------------------------------------------------
+  lsrkStep_ComputeNewDomEig:
+
+  This routine computes new dom_eig and returns SUN_SUCCESS.
+  ---------------------------------------------------------------*/
+
+int lsrkStep_ComputeNewDomEig(ARKodeMem ark_mem, ARKodeLSRKStepMem step_mem)
+{
+  int retval = SUN_SUCCESS;
+
+  retval = step_mem->dom_eig_fn(ark_mem->tn, ark_mem->ycur, ark_mem->fn,
+                                &step_mem->lambdaR, &step_mem->lambdaI,
+                                ark_mem->user_data, ark_mem->tempv1,
+                                ark_mem->tempv2, ark_mem->tempv3);
+  step_mem->dom_eig_num_evals++;
+  if (retval != ARK_SUCCESS)
+  {
+    arkProcessError(ark_mem, ARK_DOMEIG_FAIL, __LINE__, __func__, __FILE__,
+                    "Unable to estimate the dominant eigenvalue");
+    return ARK_DOMEIG_FAIL;
+  }
+
+  if (step_mem->lambdaR * ark_mem->h > ZERO)
+  {
+    arkProcessError(NULL, ARK_DOMEIG_FAIL, __LINE__, __func__, __FILE__,
+                    "lambdaR*h must be nonpositive");
+    return ARK_DOMEIG_FAIL;
+  }
+  else if (step_mem->lambdaR == SUN_RCONST(0.0) &&
+           SUNRabs(step_mem->lambdaI) > SUN_RCONST(0.0))
+  {
+    arkProcessError(NULL, ARK_DOMEIG_FAIL, __LINE__, __func__, __FILE__,
+                    "DomEig cannot be purely imaginary");
+    return ARK_DOMEIG_FAIL;
+  }
+
+  step_mem->lambdaR *= step_mem->dom_eig_safety;
+  step_mem->lambdaI *= step_mem->dom_eig_safety;
+  step_mem->spectral_radius =
+    SUNRsqrt(SUNSQR(step_mem->lambdaR) + SUNSQR(step_mem->lambdaI));
```

Comment:
That could be nice, but it might also require all of the other similar computations to be adjusted everywhere else. So, I think we should consider this in a separate PR to address all such occurrences.

Reply:
I'd defer this for later, since the current RKC and RKL methods are inappropriate when the eigenvalue has large imaginary part.  When we add RKG methods, it would be good to include `SUNRhypot`.